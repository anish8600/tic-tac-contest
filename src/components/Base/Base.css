.base-container{
    margin-left: 5px;
}