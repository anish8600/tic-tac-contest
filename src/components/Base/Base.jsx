import React from 'react';
import PropTypes from 'prop-types';
import TicTac from '../TicTac/TicTac';

class Base extends React.Component{
  formMatrix(i){
    return (
      <TicTac value={this.props.baseArray[i]} onClick={() => this.props.onClick(i)}/>
    );
  }
  
  render() {
      return (
        <div>
          <div className="base">
            {this.formMatrix(0)}
            {this.formMatrix(1)}
            {this.formMatrix(2)}
          </div>
          <div className="base">
            {this.formMatrix(3)}
            {this.formMatrix(4)}
            {this.formMatrix(5)}
          </div>
          <div className="base">
            {this.formMatrix(6)}
            {this.formMatrix(7)}
            {this.formMatrix(8)}
          </div>
        </div>
      );
    }
}

Base.propTypes = { baseArray: PropTypes.array.isRequired, onClick: PropTypes.func.isRequired};

export default Base;