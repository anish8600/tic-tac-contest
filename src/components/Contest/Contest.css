.contest-container, .contest-base {
    margin-left: 6em;
  }