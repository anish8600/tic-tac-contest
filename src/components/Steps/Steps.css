.step{
    border-radius: 3em;
  }