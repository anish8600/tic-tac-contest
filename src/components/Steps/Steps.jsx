import React from 'react';

const Steps = () => <li></li>

export default Steps;