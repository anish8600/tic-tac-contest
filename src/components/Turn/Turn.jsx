import React from 'react';
import { TIC_TAC_CONSTANTS, MAX_ATTEMPTS } from '../../helper/tictacHelper';
import PropTypes from 'prop-types';

const Turn = (props) => {
    let turnIndicator;
    if (props.champion) {
      turnIndicator = TIC_TAC_CONSTANTS.WINNER_MSG_1 + props.champion + TIC_TAC_CONSTANTS.WINNER_MSG_2;
    } else if(props.steps < MAX_ATTEMPTS){
      turnIndicator = TIC_TAC_CONSTANTS.TURN_MSG + (props.turn ? TIC_TAC_CONSTANTS.X_SYMBOL : TIC_TAC_CONSTANTS.O_SYMBOL);
    } else{
      turnIndicator = TIC_TAC_CONSTANTS.TIE_MSG;
    }
    return <h2>{turnIndicator}</h2>;
}

Turn.propTypes = { champion: PropTypes.string.isRequired, 
                   turn: PropTypes.string.isRequired,
                   steps: PropTypes.number.isRequired
                 };

export default Turn;