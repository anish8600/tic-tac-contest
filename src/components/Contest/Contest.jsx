import React from 'react';
import Base from '../Base/Base';
import Turn from '../Turn/Turn';
import Steps from '../Steps/Steps';
import { TIC_TAC_SOLUTION, TIC_TAC_CONSTANTS } from '../../helper/tictacHelper';
import './Contest.css';

function getSolution() {
    console.log('get solution');
}

class Contest extends React.Component {
  constructor(props){
    super(props);
    this.state = {
      chances: [{ baseArray: Array(9).fill(null) }],
      steps: 0,
      nextX: true
    };
  }

  updateState = () => {
    console.log('updateState operation');
  }

  revertState = () => {
    console.log('revertState operation');
  }

  render = () => {
    return (
      <div className="contest-container">
        <div className="contest-base">
          <h1> {TIC_TAC_CONSTANTS.TIC_TAC_HEADING} </h1>
          <Base baseArray={this.state.chances} onClick={this.updateState} />
        </div>
        <div className="contest-section">
          <Turn champion={'X'} turn={this.state.nextX} steps={this.state.steps}/>
          <ul>
            <Steps chances={this.state.chances} onClick={this.revertState}/>
          </ul>
        </div>
      </div>
    );
  }
}

export default Contest;