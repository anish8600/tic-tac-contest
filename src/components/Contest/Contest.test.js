import React from 'react';
import { shallow, mount} from 'enzyme';
import Contest from './Contest';
import Base from '../Base/Base';
import Turn from '../Turn/Turn';
import Steps from '../Steps/Steps';

describe('Contest', () => {
  let wrapper;
  beforeEach(() => wrapper = shallow(<Contest />));

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  });
  
  it('should render a <div />', () => {
    expect(wrapper.find('div').length).toEqual(3);
  });

  xit('should render the Base Component', () => {
    expect(wrapper.containsMatchingElement(<Base baseArray={wrapper.instance().state.chances} onClick={wrapper.instance().updateState}/>)).toEqual(true);
  });

  xit('should render the Turn Component', () => {
    expect(wrapper.containsMatchingElement(<Turn champion={'X'} turn={wrapper.instance().state.nextX} steps={wrapper.instance().state.steps} />)).toEqual(true);
  });

  xit('should render the Steps Component', () => {
    expect(wrapper.containsMatchingElement(<Steps chances={wrapper.instance().state.chances} onClick={wrapper.instance().revertState}/>)).toEqual(true);
  });

  it('should have initial state', () => {
    expect(wrapper.instance().state.chances).toEqual([ {baseArray:Array(9).fill(null)}]);
    expect(wrapper.instance().state.steps).toEqual(0);
    expect(wrapper.instance().state.nextX).toEqual(true);
  });

  it('should display correct messages in content section- Your turn, Winner', () => {
    const wrapper = mount(<Contest/>)
    const firstPlayer = wrapper.find('div.contest-section').children().first().text()
    expect(firstPlayer).toEqual('Your turn: X')

    const button = wrapper.find('button.tictac').first()
    button.simulate('click')
    const secondPlayer = wrapper.find('div.contest-section').children().first().text()
    expect(secondPlayer).toEqual('Your turn: O')

    const turn2 = wrapper.find('button.tictac').at(1)
    turn2.simulate('click')

    const turn3 = wrapper.find('button.tictac').at(4)
    turn3.simulate('click')

    const turn4 = wrapper.find('button.tictac').at(5)
    turn4.simulate('click')

    const turn5 = wrapper.find('button.tictac').at(8)
    turn5.simulate('click')

    const champion = wrapper.find('div.contest-section').children().first().text()
    expect(champion).toEqual('Congratulations X.You are a Winner!!!!!!!!')
  });
});