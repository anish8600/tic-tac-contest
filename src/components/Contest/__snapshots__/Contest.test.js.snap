// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contest should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Contest />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="contest-base"
>
          <h1>
                     
                    TIC TAC TOE CONTEST
                     
          </h1>
          <Base
                    baseArray={
                              Array [
                                        null,
                                        null,
                                        null,
                                        null,
                                        null,
                                        null,
                                        null,
                                        null,
                                        null,
                                      ]
                    }
                    onClick={[Function]}
          />
</div>,
        <div
          className="contest-section"
>
          <Turn
                    champion={null}
                    steps={0}
                    turn={true}
          />
          <ul>
                    <Steps
                              chances={
                                        Array [
                                                  Object {
                                                    "baseArray": Array [
                                                      null,
                                                      null,
                                                      null,
                                                      null,
                                                      null,
                                                      null,
                                                      null,
                                                      null,
                                                      null,
                                                    ],
                                                  },
                                                ]
                              }
                              onClick={[Function]}
                    />
          </ul>
</div>,
      ],
      "className": "contest-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
               
              TIC TAC TOE CONTEST
               
</h1>,
            <Base
              baseArray={
                            Array [
                                          null,
                                          null,
                                          null,
                                          null,
                                          null,
                                          null,
                                          null,
                                          null,
                                          null,
                                        ]
              }
              onClick={[Function]}
/>,
          ],
          "className": "contest-base",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                "TIC TAC TOE CONTEST",
                " ",
              ],
            },
            "ref": null,
            "rendered": Array [
              " ",
              "TIC TAC TOE CONTEST",
              " ",
            ],
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "baseArray": Array [
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
              ],
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Turn
              champion={null}
              steps={0}
              turn={true}
/>,
            <ul>
              <Steps
                            chances={
                                          Array [
                                                        Object {
                                                          "baseArray": Array [
                                                            null,
                                                            null,
                                                            null,
                                                            null,
                                                            null,
                                                            null,
                                                            null,
                                                            null,
                                                            null,
                                                          ],
                                                        },
                                                      ]
                            }
                            onClick={[Function]}
              />
</ul>,
          ],
          "className": "contest-section",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "champion": null,
              "steps": 0,
              "turn": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Steps
                chances={
                                Array [
                                                Object {
                                                  "baseArray": Array [
                                                    null,
                                                    null,
                                                    null,
                                                    null,
                                                    null,
                                                    null,
                                                    null,
                                                    null,
                                                    null,
                                                  ],
                                                },
                                              ]
                }
                onClick={[Function]}
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "chances": Array [
                  Object {
                    "baseArray": Array [
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                      null,
                    ],
                  },
                ],
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "ul",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="contest-base"
>
            <h1>
                         
                        TIC TAC TOE CONTEST
                         
            </h1>
            <Base
                        baseArray={
                                    Array [
                                                null,
                                                null,
                                                null,
                                                null,
                                                null,
                                                null,
                                                null,
                                                null,
                                                null,
                                              ]
                        }
                        onClick={[Function]}
            />
</div>,
          <div
            className="contest-section"
>
            <Turn
                        champion={null}
                        steps={0}
                        turn={true}
            />
            <ul>
                        <Steps
                                    chances={
                                                Array [
                                                            Object {
                                                              "baseArray": Array [
                                                                null,
                                                                null,
                                                                null,
                                                                null,
                                                                null,
                                                                null,
                                                                null,
                                                                null,
                                                                null,
                                                              ],
                                                            },
                                                          ]
                                    }
                                    onClick={[Function]}
                        />
            </ul>
</div>,
        ],
        "className": "contest-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                 
                TIC TAC TOE CONTEST
                 
</h1>,
              <Base
                baseArray={
                                Array [
                                                null,
                                                null,
                                                null,
                                                null,
                                                null,
                                                null,
                                                null,
                                                null,
                                                null,
                                              ]
                }
                onClick={[Function]}
/>,
            ],
            "className": "contest-base",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  "TIC TAC TOE CONTEST",
                  " ",
                ],
              },
              "ref": null,
              "rendered": Array [
                " ",
                "TIC TAC TOE CONTEST",
                " ",
              ],
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "baseArray": Array [
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                ],
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Turn
                champion={null}
                steps={0}
                turn={true}
/>,
              <ul>
                <Steps
                                chances={
                                                Array [
                                                                Object {
                                                                  "baseArray": Array [
                                                                    null,
                                                                    null,
                                                                    null,
                                                                    null,
                                                                    null,
                                                                    null,
                                                                    null,
                                                                    null,
                                                                    null,
                                                                  ],
                                                                },
                                                              ]
                                }
                                onClick={[Function]}
                />
</ul>,
            ],
            "className": "contest-section",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "champion": null,
                "steps": 0,
                "turn": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Steps
                  chances={
                                    Array [
                                                      Object {
                                                        "baseArray": Array [
                                                          null,
                                                          null,
                                                          null,
                                                          null,
                                                          null,
                                                          null,
                                                          null,
                                                          null,
                                                          null,
                                                        ],
                                                      },
                                                    ]
                  }
                  onClick={[Function]}
/>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "chances": Array [
                    Object {
                      "baseArray": Array [
                        null,
                        null,
                        null,
                        null,
                        null,
                        null,
                        null,
                        null,
                        null,
                      ],
                    },
                  ],
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "ul",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
